### PR TITLE
[ADD] fuzzySearch

### PR DESCRIPTION
**Pull request by kriadmin**
## Description
Add fuzzySearch to implement fuzzy search. This has many use-cases. It will help beginners easily implement search functions on their website. I would also recommend doing curry on it so that if the second argument is array then filter the array with only those which returns true when doing the fuzzy search with the first one. 

## What does your PR belong to?
- [ ] Website
- [x] Snippets
- [ ] General / Things regarding the repository (like CI Integration)
- [ ] Tests

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (non-breaking improvement of a snippet)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have checked that the changes are working properly
- [x] I have checked that there isn't any PR doing the same
- [x] I have read the **CONTRIBUTING** document.

### CODE REVIEW

File: tag_database [hunk 1]
```diff
@@ -283,4 +284,4 @@ xProd:array,math
 yesNo:utility,regexp
 zip:array
 zipObject:array,object
-zipWith:array,function,advanced
+zipWith:uncategorized
```
Comment (Chalarangelo) [inline tag_database]: Why were `zipWith`'s tags removed?

Comment (kriadmin) [reply tag_database]: Sorry I didn't know it was removed

**Comment by Chalarangelo** [comment]
I quite like the idea of a fuzzy search, however there are three major considerations we need to discuss before finalizing the PR:

1. Is this snippet actually appropriate or does it fall under the category of algorithmic implementations that might not be a good fit for the main repo? I think it's adequate and simple enough, wouldn't harm to have it on the main list.
2. Is this a good/complete implementation of a fuzzy search algorithm? I'm not entirely certain about this, I will check some theory, but I think there might be some improvements to be made.
3. Is the code well-written? Now, technically speaking, the code works just fine, but there are a lot of improvements that can be made to it before merging, so I will take some time to improve it a little bit before we finally accept the PR.

Overall, I think this is a snippet with a lot of potential, so it'll most likely make it on the main list, unless some of the other maintainers have any objections.

File: snippets/fuzzySearch.md [hunk 2]
```diff
@@ -0,0 +1,33 @@
+### fuzzySearch
+
+Determines if the `patrn` matches with `str`
+
+Loops through `str` and determines if it contains all characters of `patrn` and in the correct order. Both the strings are converted to lower case.
+
+Taken from [here](https://github.com/forrestthewoods/lib_fts/blob/80f3f8c52db53428247e741b9efe2cde9667050c/code/fts_fuzzy_match.js#L18).
+``` js
+fuzzySearch = (patrn, str) => {
+    const pattern = patrn;
```
Comment (skatcat31) [inline snippets/fuzzySearch.md]: Why is there a re-declaration?

File: snippets/fuzzySearch.md [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    const string = str; 
+    let patternIdx = 0;
+    let strIdx = 0;
+    let patternLength = pattern.length;
```
Comment (skatcat31) [inline snippets/fuzzySearch.md]: un-needed declarations

File: snippets/fuzzySearch.md (hunk 2 above)
Comment (skatcat31) [reply snippets/fuzzySearch.md]: By shadow assigning to an equivilant value we can skip declaring the length property statically so we can have fewer lines of code
```js
const fuzzySearch = (pattern, string) => {
  pattern = String(pattern);
  string = String(string);
```

File: snippets/fuzzySearch.md [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+Determines if the `patrn` matches with `str`
+
+Loops through `str` and determines if it contains all characters of `patrn` and in the correct order. Both the strings are converted to lower case.
+
+Taken from [here](https://github.com/forrestthewoods/lib_fts/blob/80f3f8c52db53428247e741b9efe2cde9667050c/code/fts_fuzzy_match.js#L18).
+``` js
+fuzzySearch = (patrn, str) => {
+    const pattern = patrn;
+    const string = str; 
+    let patternIdx = 0;
```
Comment (skatcat31) [inline snippets/fuzzySearch.md]: If we do a for loop we can avoid declaring above

File: snippets/fuzzySearch.md [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    let strIdx = 0;
+    let patternLength = pattern.length;
+    let strLength = string.length;
+
+    while (patternIdx !== patternLength && strIdx !== strLength) {
```
Comment (skatcat31) [inline snippets/fuzzySearch.md]: There are other looping choices that might be a better fit

PS: This specifies that it MUST be in order of pattern. There is also a way to do this same search as unordered, but I'm not sure that adhere's to what a fuzzy search is.

File: snippets/fuzzySearch.md (hunk 2 above)
Comment (kriadmin) [reply snippets/fuzzySearch.md]: Re-declaration is needed to avoid changing the values

File: snippets/fuzzySearch.md (hunk 5 above)
Comment (kriadmin) [reply snippets/fuzzySearch.md]: I kept it short to adhere to the contribution guidelines. This is not a complete fuzzy search implementation.

File: snippets/fuzzySearch.md [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+Loops through `str` and determines if it contains all characters of `patrn` and in the correct order. Both the strings are converted to lower case.
+
+Taken from [here](https://github.com/forrestthewoods/lib_fts/blob/80f3f8c52db53428247e741b9efe2cde9667050c/code/fts_fuzzy_match.js#L18).
+``` js
+fuzzySearch = (pattern, string) => { 
+    let patternLength = pattern.length;
+    let strLength = string.length;
+
+    for (var patternIdx = 0, strIdx = 0;patternIdx !== patternLength && strIdx !== strLength;strIdx++){
```
Comment (Chalarangelo) [inline snippets/fuzzySearch.md]: I'm pretty certain this whole loop could just be a single `Array.reduce()` with some creative trickery. I'm gonna try my hand at it and let you know if it's doable and readable.

**Comment by Chalarangelo** [comment]
### Update

The whole method could be reduced (pun unintended) to a single statement, using `Array.reduce()` and spreading the string, like this:

```js
fuzzySearch = (pattern, str) =>
	[...str].reduce(
		(acc, char) =>
			char.toLowerCase() === pattern[acc].toLowerCase() ? acc + 1 : acc,
		0
	) === pattern.length
		? true
		: false;
```

I think this might be a good example of using `Array.reduce()` for something a bit non-conventional, so I vote that we convert the method to look like this. It's also significantly shorter.

File: snippets/fuzzySearch.md [hunk 7]
```diff
@@ -0,0 +1,28 @@
+### fuzzySearch
+
+Determines if the `patrn` matches with `str`
+
+Loops through `str` and determines if it contains all characters of `patrn` and in the correct order. Both the strings are converted to lower case.
+
+Taken from [here](https://github.com/forrestthewoods/lib_fts/blob/80f3f8c52db53428247e741b9efe2cde9667050c/code/fts_fuzzy_match.js#L18).
+``` js
+fuzzySearch = (pattern, string) => {
```
Comment (Chalarangelo) [inline snippets/fuzzySearch.md]: Missing a `const` keyword here.

**Comment by kriadmin** [comment]
@Chalarangelo Updated it using your reduce but i am not sure if everyone will understand it :confused: .

**Comment by Chalarangelo** [comment]
@kriadmin It's a pretty simple snippet even with `Array.reduce()`. All it does is iterate over the letters of the string, checking if the letter at the pattern's index is matching and increments or keeps as-is the pattern index. I don't think there's a lot to understand once someone reads how it works, it's quite straightforward.

File: snippets/fuzzySearch.md [hunk 8]
```diff
@@ -0,0 +1,19 @@
+### fuzzySearch
+
+Determines if the `patrn` matches with `str`
```
Comment (skatcat31) [inline snippets/fuzzySearch.md]: Needs to be updated

`Determines if the provided pattern is matched in order within a string`

File: snippets/fuzzySearch.md [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+
+Loops through `str` and determines if it contains all characters of `patrn` and in the correct order. Both the strings are converted to lower case.
```
Comment (skatcat31) [inline snippets/fuzzySearch.md]: patrn was redone to pattern with no shadow redecleration

File: snippets/fuzzySearch.md [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+
+Taken from [here](https://github.com/forrestthewoods/lib_fts/blob/80f3f8c52db53428247e741b9efe2cde9667050c/code/fts_fuzzy_match.js#L18).
```
Comment (skatcat31) [inline snippets/fuzzySearch.md]: It's not longer taken, but adapted from

File: snippets/fuzzySearch.md [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+``` js
+const fuzzySearch = (pattern, str) =>
+	[...str].reduce(
+		(acc, char) => char.toLowerCase() === (pattern[acc]  || '').toLowerCase() ? acc + 1 : acc, 0
```
Comment (skatcat31) [inline snippets/fuzzySearch.md]: May be more readable as `matchIndex` instead of `acc`

File: test/fuzzySearch/fuzzySearch.js [hunk 12]
```diff
@@ -0,0 +1,19 @@
+fuzzySearch = (patrn, str) => {
```
Comment (skatcat31) [inline test/fuzzySearch/fuzzySearch.js]: This file needs to be removed

Comment (skatcat31) [reply test/fuzzySearch/fuzzySearch.js]: or regenerated rather

File: snippets/fuzzySearch.md [hunk 13]
```diff
@@ -0,0 +1,19 @@
+### fuzzySearch
```
Comment (kingdavidmartins) [inline snippets/fuzzySearch.md]: I believe we definitely need to change  the snippet name from `fuzzySearch` to something else like the following 

`lazySearch`
`lazyMatch`
`isSimilar`

To be honest I am currently indifferent as to the exact name & okay with whatever we agree on

File: test/fuzzySearch/fuzzySearch.js (hunk 12 above)
Comment (kingdavidmartins) [reply test/fuzzySearch/fuzzySearch.js]: Agreed! After we agree on the final snippet. Please run `npm run tester` so generated files can reflect most recent changes

**Comment by Chalarangelo** [comment]
I believe that this is a snippet that might better fit the archive base on @kingdavidmartins's input as well as qualifying for a Medium article to explain the methods and functionality more in depth than a snippet can.

**Comment by kriadmin** [comment]
@kingdavidmartins @skatcat31 @Chalarangelo So i shall update the snippets based on the recommended changes. But can you all finalize the name? I am sorry i can't be any help in that.

**Comment by skatcat31** [comment]
Sure we can finalize the name. We just want you to get credit for the code

**Comment by fejes713** [comment]
I am a bit late to the party but I like the snippet overall. I totally agree with @Chalarangelo and @kingdavidmartins, like they said this snippet should go to archive base. Speaking about the name, I think that `lazySearch` should be fine. However, I am open to all the ideas.

**Comment by kriadmin** [comment]
I have updated the snippet as requested.

**Comment by skatcat31** [comment]
Updated the test file, but have not rerun tests as I am not at home. If someone wants to rerun tests it should be good. Now all we need is a new name.

I'm voting for `isSimilarString` or `isSimilar`

**Comment by kingdavidmartins** [comment]
I am also leaning towards `isSimilar`

File: test/fuzzySearch/fuzzySearch.test.js [hunk 14]
```diff
@@ -0,0 +1,15 @@
+const test = require('tape');
```
Comment (kingdavidmartins) [inline test/fuzzySearch/fuzzySearch.test.js]: Please run `npm run tester`

File: test/fuzzySearch/fuzzySearch.js [hunk 15]
```diff
@@ -0,0 +1,5 @@
+const fuzzySearch = (pattern, str) =>
```
Comment (kingdavidmartins) [inline test/fuzzySearch/fuzzySearch.js]: Please run `npm run tester`

**Pull request merged**